src/java/org/apache/cassandra/utils/concurrent/RateLimiter.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils.concurrent;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.cassandra.utils.ExpMovingAverage;
+
+/**
+ * A RateLimiter very much like {@link com.google.common.util.concurrent.RateLimiter} but this one
+ * also measures and exposes the actual rate.
+ */
+@SuppressWarnings("UnstableApiUsage")
+public class RateLimiter
```

Comment:
It would be better to port over [CASSANDRA-13890](https://issues.apache.org/jira/browse/CASSANDRA-13890) which introduces a compaction metric for the current compaction rate (with a few aggregations, e.g. 1-minute rate).